app/models/metadatum.rb

```diff
@@ -90,25 +111,113 @@ def check_and_set_string_type
     end
   end
 
-  # Add a new Metadatum entry to the sample
-  def self.add_to_sample(sample, key, raw_value)
-    key = key.to_sym
-
-    # Create the entry
-    m = Metadatum.new
-    m.key = key
-    m.data_type = KEY_TO_TYPE[key]
-    if KEY_TO_TYPE[key] == STRING_TYPE
-      m.text_raw_value = raw_value
-    elsif KEY_TO_TYPE[key] == NUMBER_TYPE
-      m.number_raw_value = raw_value
+  # Set value based on data type
+  def edit_value(val)
+    if data_type == "string"
+      self.text_raw_value = val
+    elsif data_type == "number"
+      self.number_raw_value = val
     end
-    # *_validated_value field is set in the set_validated_values validator.
-    m.sample = sample
-    m.save!
+    # Note: *_validated_value field is set in the set_validated_values
+    # validator.
   end
 
-  def str_to_basic_chars(res)
+  def self.str_to_basic_chars(res)
     res.downcase.gsub(/[^0-9A-Za-z]/, '')
   end
+
+  # Load bulk metadata from a CSV file from S3
+  def self.bulk_load_from_s3_csv(path)
```

Comment:
So if I try to upload a 20-column metadata file for project "Uganda - all - 2" (214 samples), this will make 4280 database queries?

Reply:
I think this would be a good time to improve on that

Reply:
the previous implementation at least updated all the columns in one go for each row ;). But better would be: do 1 query to retrieve the data needed to make the validations. Iterate through the data and enforce the validations. Then do 1 query to write all the data. It will require some reorganization of the code, but if we ever seriously want to use metadata adding/editing functionality, it will be indispensable.

Reply:
There are more queries in the process, like loading the project per row and and the sample per row that can be optimized by grouping everything.

Reply:
- Rough benchmark on staging, inserting 5000 Metadata will take about 15-30 seconds. So to enable a better user experience for the sake of uploading in the frontend I'll try to get that down with bulk inserts.
- I'd like to keep things in ActiveRecord though for keeping things the Rails way as much as possible.. the performance gain will be from not having to do the roundtrip transaction/commit/etc for each Metadata.